python/lsst/ts/scheduler/model.py

```diff
@@ -0,0 +1,995 @@
+# This file is part of ts_scheduler
+#
+# Developed for the Rubin Observatory Telescope and Site Systems.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["Model"]
+
+import asyncio
+import functools
+import inspect
+import io
+import logging
+import pathlib
+import typing
+import urllib.request
+from importlib import import_module
+
+import numpy as np
+from lsst.ts.astrosky.model import AstronomicalSkyModel
+from lsst.ts.dateloc import ObservatoryLocation
+from lsst.ts.idl.enums import Script
+from lsst.ts.observatory.model import ObservatoryModel, ObservatoryState
+from rubin_sim.site_models.cloudModel import CloudModel
+from rubin_sim.site_models.downtimeModel import DowntimeModel
+from rubin_sim.site_models.seeingModel import SeeingModel
+
+from lsst.ts import utils
+
+from .driver import Driver
+from .driver.driver_target import DriverTarget
+from .driver.survey_topology import SurveyTopology
+from .telemetry_stream_handler import TelemetryStreamHandler
+from .utils.csc_utils import NonFinalStates, is_uri, is_valid_efd_query
+from .utils.exceptions import UpdateTelemetryError
+from .utils.scheduled_targets_info import ScheduledTargetsInfo
+
+
+class Model:
+    """Scheduler model class.
+
+    This class implements the business logic of the Scheduler. The CSC
+    instantiate the model and offload the operations to it, which allows for
+    better separation of concerns and cleaner code on the CSC part.
+
+    Attributes
+    ----------
+    telemetry_stream_handler : `TelemetryStreamHandler`
+        Object to handle telemetry stream.
+    models : `dict`[`str`, `Any`]
+        Dictionary to store the scheduler models.
+    raw_telemetry : `dict`[`str`, `Any`]
+        Dictionary to store raw telemetry.
+    driver : `Driver`
+        Lock for the scheduler detailed state.
+    """
+
+    def __init__(self, log) -> None:
+
+        self.log = log.getChild(type(self).__name__)
+
+        self.telemetry_stream_handler = None
+
+        # How far to step into the future when there's not targets in seconds
+        self.time_delta_no_target = 30.0
```

Comment:
Please extract to a constant outside of this class.

Reply:
I though about that, but I am still deciding whether this is a constant or not. I think I may allow this to be configurable in the future.